File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    private int myBinarySearch(int[] nums, int low, int high, int target)
    {
        while(low <= high)
        {
            int mid = (high + low) / 2;
            
            if(nums[mid] == target)
                return mid;
            
            else if(nums[mid] < target)
                low = mid + 1;
            
            else
                high = mid - 1;
        }
        
        return -1;
    }
    
    public int search(int[] nums, int target) {
        int low = 0, high = nums.length - 1, minIndex = -1;
        
        while(low < high && nums[low] > nums[high])
        {
            int mid = (high + low) / 2;
            
            if(nums[mid + 1] < nums[mid])
            {
                minIndex = mid + 1;
                break;
            }

            if(nums[mid] > nums[low])
                low = mid + 1;

            else
                high = mid;
        }
        
        
        if(minIndex == -1)
            minIndex = low;
        
        int search1 = myBinarySearch(nums, 0, minIndex - 1, target);
        int search2 = myBinarySearch(nums, minIndex, nums.length - 1, target);
        if(search1 != -1)
            return search1;

        if(search2 != -1)
            return search2;

        return -1;
    }
}

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int n = nums.size();
        int low = 0, high = n - 1;
        
        while (low <= high) 
        {
            int mid = midpoint(low, high);
            
            if (nums[mid] == target)
                return mid;
            
            else if (nums[mid] >= nums[low]) // Left side is sorted.
            {
                if (target >= nums[low] and target < nums[mid]) 
                    high = mid - 1;
                
                else 
                    low = mid + 1;
            }
            
            else // Right side is sorted.
            {
                if (target <= nums[high] and target > nums[mid])
                    low = mid + 1;
                
                else
                    high = mid - 1;
            }
        }
        
        return -1;
    }
};